tensorflow/java/src/gen/cc/op_generator.cc

```diff
@@ -14,53 +14,429 @@ limitations under the License.
+      .attributes(string("value = \"op_generator\", date = \"") + date + "\""));
+  if (endpoint.deprecated()) {
+    op_class.add_annotation(Annotation::Create("Deprecated"));
+    string explanation;
+    if (!op.endpoints().front().deprecated()) {
+      explanation = "use {@link " +
+          ClassOf(op.endpoints().front(), base_package).full_name()
+          + "} instead";
+    } else {
+      explanation = op.deprecation_explanation();
+    }
+    op_javadoc.add_tag("deprecated", explanation);
+  }
+  if (!op.hidden()) {
+    // expose the op in the Ops Graph API only if it is visible
+    op_class.add_annotation(
+        Annotation::Create("Operator", "org.tensorflow.op.annotation")
+          .attributes("group = \"" + endpoint.package() + "\""));
+  }
+  // create op class file
+  string op_dir = io::JoinPath(output_dir,
+      str_util::StringReplace(op_class.package(), ".", "/", true));
+  if (!env->FileExists(op_dir).ok()) {
+    TF_CHECK_OK(Env::Default()->RecursivelyCreateDir(op_dir));
+  }
+  std::unique_ptr<tensorflow::WritableFile> op_file;
+  TF_CHECK_OK(env->NewWritableFile(
+      io::JoinPath(op_dir, op_class.name() + ".java"), &op_file));
+
+  // render endpoint source code
+  SourceFileWriter writer(op_file.get());
+  std::list<Type> dependencies;
+  CollectOpDependencies(op, mode, &dependencies);
+  writer.WriteFromFile(kLicenseSnippet)
+      .EndLine()
+      .BeginType(op_class, PUBLIC|FINAL, &dependencies, &op_javadoc);
+  if (!op.optional_attributes().empty()) {
+    RenderOptionsClass(op, op_class, &writer);
+  }
+  RenderFactoryMethods(op, op_class, &writer);
+  RenderGettersAndSetters(op, &writer);
+  if (mode != DEFAULT) {
+    RenderInterfaceImpl(op, mode, &writer);
+  }
+  writer.EndLine();
+  for (const ArgumentSpec& output : op.outputs()) {
+    writer.WriteField(output.var(), PRIVATE);
+  }
+  RenderConstructor(op, op_class, &writer);
+  writer.EndType();
+}
+
+}  // namespace
+
+Status OpGenerator::Run(const OpList& op_list, const string& base_package,
+    const string& output_dir) {
+  ApiDefMap api_map(op_list);
+  if (!api_dirs_.empty()) {
+    // Only load api files that correspond to the requested "op_list"
+    for (const auto& op : op_list.op()) {
+      for (const auto& api_def_dir : api_dirs_) {
+        const std::string api_def_file_pattern =
+            io::JoinPath(api_def_dir, "api_def_" + op.name() + ".pbtxt");
+        if (env_->FileExists(api_def_file_pattern).ok()) {
+          TF_CHECK_OK(api_map.LoadFile(env_, api_def_file_pattern));
+        }
+      }
+    }
+  }
+  api_map.UpdateDocs();
+  time_t now;
+  time(&now);
+  for (const auto& op_def : op_list.op()) {
+    const ApiDef* api_def = api_map.GetApiDef(op_def.name());
+    if (api_def->visibility() != ApiDef::SKIP) {
+      OpSpec op(OpSpec::Create(op_def, *api_def));
+      for (const EndpointSpec& endpoint : op.endpoints()) {
+        GenerateOp(op, endpoint, base_package, output_dir, env_, gmtime(&now));
```

Comment:
I'd recommend against including the timestamp as an annotation in the generated files, that leads to spurious differences between identical builds.